src/hera/workflows/runner.py

```diff
@@ -122,7 +123,66 @@ def _map_keys(function: Callable, kwargs: dict) -> dict:
     return mapped_kwargs
 
 
-def _runner(entrypoint: str, kwargs_list: Any) -> str:
+def _save_outputs(
+    function_results: Union[Tuple[Any], Any], output_destinations: List[Tuple[type, Union[Parameter, Artifact]]]
+) -> None:
+    """Save the results of the function to the specified output destinations.
+
+    The results are matched with the specified outputs and saved using the schema:
+    global_config.experimental_features["hera_output_directory"]/parameters/<name>
+    global_config.experimental_features["hera_output_directory"]/artifacts/<name>
+
+    If the artifact path is specified, that is used instead.
+    """
+    if not isinstance(function_results, tuple):
+        function_results = [function_results]
+    if len(function_results) != len(output_destinations):
+        raise ValueError("The number of outputs does not match the annotation")
+
+    output_dir = Path(
+        str(global_config.experimental_features["hera_output_directory"])
+        if global_config.experimental_features["hera_output_directory"]
+        else "hera/outputs"
+    )
+
+    for res, dest in zip(function_results, output_destinations):
+        if not isinstance(res, dest[0]):
+            raise ValueError(
+                f"The type of output `{dest[1].name}`, `{type(res)}` does not match the annotated type `{dest[0]}`"
+            )
+        if not dest[1].name:
+            raise ValueError("The name was not provided for one of the outputs.")
+
+        if isinstance(dest[1], Parameter):
+            path = output_dir / f"parameters/{cast(str, dest[1].name)}"
+        elif isinstance(dest[1], Artifact):
+            if dest[1].path:
+                path = Path(dest[1].path)
+            else:
+                path = output_dir / f"artifacts/{cast(str, dest[1].name)}"
+        _write_to_path(path, res)
+
+
+def _extract_output_annotations(function: Callable) -> Optional[List]:
+    output = []
+    if get_origin(inspect.signature(function).return_annotation) is Annotated:
+        output.append(get_args(inspect.signature(function).return_annotation))
```

Comment:
Can we make variables like

`origin_type = get_origin(inspect.signature(function).return_annotation)`

and

`annotation_args = get_args(inspect.signature(function).return_annotation)`

to make this a bit easier to read?